Components/Optimizers/PreconditionedGradientDescent/itkStochasticPreconditionedGradientDescentOptimizer.h

```diff
@@ -0,0 +1,166 @@
+/*=========================================================================
+ *
+ *  Copyright UMC Utrecht and contributors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *=========================================================================*/
+#ifndef __itkStochasticPreconditionedGradientDescentOptimizer_h
+#define __itkStochasticPreconditionedGradientDescentOptimizer_h
+
+#include "itkPreconditionedGradientDescentOptimizer.h"
+
+namespace itk
+{
+/**
+ * \class StochasticPreconditionedGradientDescentOptimizer
+ * \brief This class implements a gradient descent optimizer with a decaying gain and preconditioning
+ *
+ * If \f$C(x)\f$ is a cost function that has to be minimized, the following iterative
+ * algorithm is used to find the optimal parameters \f$x\f$:
+ *
+ *     \f[ x(k+1) = x(k) - a(k) P dC/dx \f]
+ *
+ * The gain \f$a(k)\f$ at each iteration \f$k\f$ is defined by:
+ *
+ *     \f[ a(k) =  a / (A + k + 1)^alpha \f].
+ *
+ * It is very suitable to be used in combination with a stochastic estimate
+ * of the gradient \f$dC/dx\f$. For example, in image registration problems it is
+ * often advantageous to compute the metric derivative (\f$dC/dx\f$) on a new set
+ * of randomly selected image samples in each iteration. You may set the parameter
+ * \c NewSamplesEveryIteration to \c "true" to achieve this effect.
+ * For more information on this strategy, you may have a look at:
+ *
```

Comment:
Qiao Y, Lelieveldt BP, Staring M. An efficient preconditioner for stochastic gradient descent optimization of image registration. IEEE transactions on medical imaging. 2019 Feb 11.